### ISSUE TITLE
"Persistence directory already locked by this process" exception upon app redeployment despite prior closure of CacheManager in ServletContextListener

### ISSUE THREAD

**Issue by haqer1** [ehcache/ehcache3#3007]
1) CacheManager is closed as follows (assuming context is destroyed when re-deploying the app):
```java
@WebListener
public class EhCacheManagerCloser implements ServletContextListener {
	public EhCacheManagerCloser() { }

	@Override
	public void contextDestroyed(ServletContextEvent arg0) {
		CacheManager cm = AbstractJCacheProvider.getInstance().getCacheManager();
		if (cm instanceof EhCacheJSR107Provider)
			cm.close();
	}

	@Override
	public void contextInitialized(ServletContextEvent arg0) { }
}
```

2) Exception thrown is as follows:
Caused by: org.ehcache.StateTransitionException: Persistence directory already locked by this process: /some/path/ehcache
        at org.ehcache.core.StatusTransitioner$Transition.failed(StatusTransitioner.java:235)
        at org.ehcache.core.EhcacheManager.init(EhcacheManager.java:605)
        at org.ehcache.jsr107.EhcacheCachingProvider.createCacheManager(EhcacheCachingProvider.java:159)
        at org.ehcache.jsr107.EhcacheCachingProvider.getCacheManager(EhcacheCachingProvider.java:134)
        at org.ehcache.jsr107.EhcacheCachingProvider.getCacheManager(EhcacheCachingProvider.java:85)
        at org.ehcache.jsr107.EhcacheCachingProvider.getCacheManager(EhcacheCachingProvider.java:194)
        at some.package.util.cache.AbstractJCacheProvider.initCacheManagerFromXMLConfig(AbstractJCacheProvider.java:65)
        at some.package.util.cache.EhCacheJSR107Provider.<init>(EhCacheJSR107Provider.java:80)
        at some.package.util.cache.EhCacheJSR107Provider.<init>(EhCacheJSR107Provider.java:72)
        at some.package.util.cache.EhCacheJSR107Provider.<clinit>(EhCacheJSR107Provider.java:34)
        ... 78 more
Caused by: java.lang.RuntimeException: Persistence directory already locked by this process: /some/path/ehcache
        at org.ehcache.impl.persistence.DefaultLocalPersistenceService.internalStart(DefaultLocalPersistenceService.java:95)
        at org.ehcache.impl.persistence.DefaultLocalPersistenceService.start(DefaultLocalPersistenceService.java:75)
        at org.ehcache.core.spi.ServiceLocator.startAllServices(ServiceLocator.java:128)
        at org.ehcache.core.EhcacheManager.init(EhcacheManager.java:577)
        ... 86 more
Caused by: java.nio.channels.OverlappingFileLockException
        at sun.nio.ch.SharedFileLockTable.checkList(FileLockTable.java:255)
        at sun.nio.ch.SharedFileLockTable.add(FileLockTable.java:152)
        at sun.nio.ch.FileChannelImpl.tryLock(FileChannelImpl.java:1107) 
        at java.nio.channels.FileChannel.tryLock(FileChannel.java:1155) 
        at org.ehcache.impl.persistence.DefaultLocalPersistenceService.internalStart(DefaultLocalPersistenceService.java:93) 
        ... 89 more 

P.S. Work-around currently used:
a. stop server
b. update app
c. re-start server
Needless to say that is sub-optimal.
P.P.S. App server: GlassFish 4

**Comment by haqer1** [comment]
Well, the following line is buggy:
```java
if (cm instanceof EhCacheJSR107Provider)
```

I commented it out, & will re-try this as time allows...

**Comment by haqer1** [comment]
Sorry for bothering: that indeed was a bug in my code. Closing the issue.

P.S. If anybody is interested, here are the main snippets w/ fix-ups.
```java
@WebListener
public class EhCacheManagerCloser implements ServletContextListener {
	private static final Logger logger = Logger.getLogger(EhCacheManagerCloser.class.getName());
	private static final String EH_107_CACHE_MANAGER_CLASS_NAME = "Eh107CacheManager";

	private static boolean isCloseable(CacheManager cm) {
		return // class isn't public, so check type by simple name:
			cm.getClass().getSimpleName().equals(EH_107_CACHE_MANAGER_CLASS_NAME);
	}

	public EhCacheManagerCloser() { }

	@Override
	public void contextDestroyed(ServletContextEvent arg0) {
		CacheManager cm = AbstractJCacheProvider.getInstance().getCacheManager();
		if (isCloseable(cm)) { // fix-up 1.1
			if (logger.isLoggable(Level.INFO))
				logger.info("Closing Eh107CacheManager...");
			cm.close();
		}
	}

	@Override
	public void contextInitialized(ServletContextEvent arg0) { }
}

// ...

public abstract class AbstractJCacheProvider {
	protected boolean useXmlBasedConfig;
// ...
	protected AbstractJCacheProvider(boolean useXmlBasedConfig) {
		this.useXmlBasedConfig = useXmlBasedConfig;
		this.cacheManager = initCacheManager();
	}

	abstract CacheManager initCacheManager();
// ...

	public CacheManager getCacheManager() {
		if (cacheManager.isClosed())
			synchronized (this) {
				if (cacheManager.isClosed())
					cacheManager = initCacheManager(); // fix-up 1.2
			}
		return cacheManager;
	}
}
// ...
```
P.P.S. Redeployments work fine now (no errors thrown). And here's some logging confirming what's happening, keywords being "removed", followed by "created".
[2022-04-11T08:55:09.266+0000] [Payara 4.1] [INFO] [] [some.package.config.EhCacheManagerCloser] [tid: _ThreadID=112 _ThreadName=admin-thread-pool::admin-listener(2)] [timeMillis: 1649667309266] [levelValue: 800] [[
  Closing Eh107CacheManager...]]

[2022-04-11T08:55:09.292+0000] [Payara 4.1] [INFO] [] [org.ehcache.core.EhcacheManager] [tid: _ThreadID=112 _ThreadName=admin-thread-pool::admin-listener(2)] [timeMillis: 1649667309292] [levelValue: 800] [[
  Cache 'someCache' removed from EhcacheManager.]]

[2022-04-11T08:55:09.308+0000] [Payara 4.1] [INFO] [] [org.ehcache.core.EhcacheManager] [tid: _ThreadID=112 _ThreadName=admin-thread-pool::admin-listener(2)] [timeMillis: 1649667309308] [levelValue: 800] [[
  Cache 'someOtherCache' removed from EhcacheManager.]]
...

[2022-04-11T08:55:15.098+0000] [Payara 4.1] [INFO] [] [some.other.package.AbstractJCacheProvider] [tid: _ThreadID=112 _ThreadName=admin-thread-pool::admin-listener(2)] [timeMillis: 1649667315098] [levelValue: 800] [[
  DEFAULT_CACHE_ALIAS: someCache]]

[2022-04-11T08:55:15.800+0000] [Payara 4.1] [INFO] [] [org.ehcache.impl.internal.store.disk.OffHeapDiskStore] [tid: _ThreadID=112 _ThreadName=admin-thread-pool::admin-listener(2)] [timeMillis: 1649667315800] [levelValue: 800] [[
  The index for data file ehcache-disk-store.data is more recent than the data file itself by 100282ms : this is harmless.]]

[2022-04-11T08:55:15.850+0000] [Payara 4.1] [INFO] [] [org.ehcache.core.EhcacheManager] [tid: _ThreadID=112 _ThreadName=admin-thread-pool::admin-listener(2)] [timeMillis: 1649667315850] [levelValue: 800] [[
  Cache 'someCache' created in EhcacheManager.]]

[2022-04-11T08:55:15.855+0000] [Payara 4.1] [INFO] [] [org.ehcache.impl.internal.store.disk.OffHeapDiskStore] [tid: _ThreadID=112 _ThreadName=admin-thread-pool::admin-listener(2)] [timeMillis: 1649667315855] [levelValue: 800] [[
  The index for data file ehcache-disk-store.data is more recent than the data file itself by 100298ms : this is harmless.]]

[2022-04-11T08:55:15.858+0000] [Payara 4.1] [INFO] [] [org.ehcache.core.EhcacheManager] [tid: _ThreadID=112 _ThreadName=admin-thread-pool::admin-listener(2)] [timeMillis: 1649667315858] [levelValue: 800] [[
  Cache 'someOtherCache' created in EhcacheManager.]]

[2022-04-11T08:55:15.922+0000] [Payara 4.1] [INFO] [] [org.ehcache.jsr107.Eh107CacheManager] [tid: _ThreadID=112 _ThreadName=admin-thread-pool::admin-listener(2)] [timeMillis: 1649667315922] [levelValue: 800] [[
  Registering Ehcache MBean javax.cache:type=CacheStatistics,CacheManager=jar.file./some/path/WEB-INF/lib/some-artifact-jar-0.9.16-SNAPSHOT.jar!/ehcache.xml,Cache=someCache]]

[2022-04-11T08:55:15.924+0000] [Payara 4.1] [INFO] [] [org.ehcache.jsr107.Eh107CacheManager] [tid: _ThreadID=112 _ThreadName=admin-thread-pool::admin-listener(2)] [timeMillis: 1649667315924] [levelValue: 800] [[
  Registering Ehcache MBean javax.cache:type=CacheStatistics,CacheManager=jar.file./some/path/WEB-INF/lib/some-artifact-jar-0.9.16-SNAPSHOT.jar!/ehcache.xml,Cache=someOtherCache]]
